lib/kino/screen.ex

```diff
@@ -0,0 +1,303 @@
+      state = fun.(data, state)
+      {:noreply, render(module, frame, client_id, state)}
+    end
+
+    defp render(module, frame, client_id, state) do
+      Kino.Frame.render(frame, module.render(state), to: client_id)
+      {module, frame, client_id, state}
+    end
+  end
+
+  defmodule Watcher do
+    @moduledoc false
+    use GenServer
+
+    def start_link(mod_frame_state) do
+      GenServer.start_link(__MODULE__, {mod_frame_state, self()})
+    end
+
+    @impl true
+    def init({{module, frame, state}, parent}) do
+      Kino.Frame.render(frame, module.render(state))
+
+      data = %{
+        module: module,
+        frame: frame,
+        state: state,
+        sup: parent,
+        children: %{}
+      }
+
+      {:ok, data, {:continue, {:init, parent}}}
+    end
+
+    @impl true
+    def handle_continue({:init, parent}, state) do
+      [_, {_id, sup, _, _}] = Supervisor.which_children(parent)
+      {:noreply, %{state | sup: sup}}
+    end
+
+    @impl true
+    def handle_info({{Kino.Screen, fun}, event}, data) do
+      if not Map.has_key?(event, :origin) do
+        raise "expected control/2 to map to an event with origin"
+      end
+
+      %{module: module, frame: frame, state: state, sup: sup, children: children} = data
+
+      children =
+        case DynamicSupervisor.start_child(sup, {Server, {module, frame, fun, event, state}}) do
+          {:ok, pid} -> Map.put(children, event.origin, pid)
+          {:error, _} -> children
+        end
+
+      {:noreply, %{data | children: children}}
+    end
+
+    def handle_info({:client_leave, client_id}, %{sup: sup, children: children} = data) do
+      {pid, children} = Map.pop(children, client_id)
+      pid && DynamicSupervisor.terminate_child(sup, pid)
+      {:noreply, %{data | children: children}}
+    end
+  end
+
+  @typedoc "The state of the screen"
+  @type state :: term()
+
+  @doc """
+  Callback invoked when the screen is initialized.
+
+  It receives the second argument given to `new/2` and
+  it must return the screen state.
+
+  This function is called in the same process that calls
+  `Kino.Screen.new/2`.
+  """
+  @callback init(state) :: {:ok, state}
+
+  @doc """
+  Callback invoked to render the screen, whenever there
+  is a control event.
+
+  It receives the state and it must return a renderable
+  output. The first time this function is called, it is
+  done within a temporary process until the user first
+  interacts with the form.
+  """
+  @callback render(state) :: term()
+
+  @doc """
+  Receives a control or an input and invokes the given 2-arity function
+  once its actions are triggered.
+  """
+  @spec control(element, (map(), state() -> state())) :: element
+        when element: Kino.Control.t() | Kino.Input.t()
+
+  def control(from, fun) when is_function(fun, 2) do
+    Kino.Control.subscribe(from, {__MODULE__, fun})
+    from
+  end
+
+  def new(module, state) when is_atom(module) do
+    {:ok, state} = module.init(state)
```

Comment:
@jonatanklosko, this may be a bit confusing.

1. We call `module.init/1` in the caller process, so we don't fall into the trap of starting a frame inside a process
2. The first `render/1` happens in the watcher process
3. All following render happens in the user-specific process

We are exposing three processes to users. Given when `module.init/1` is called today, I am thinking we should just get rid of it and you pass the state as argument to Kino.Screen.new directly. We may add it back in the future to customize the user specific process instead.